src/algebra/homology/homotopy.lean

```diff
@@ -0,0 +1,507 @@
+  from_next' (λ i j, f i j ≫ g.f j) i j = from_next' f i j ≫ g.f j :=
+begin
+  dsimp [from_next'],
+  rcases c.next i with _|⟨i',w⟩,
+  { exact zero_comp.symm, },
+  { dsimp [from_next'._match_1, hom.next._match_1],
+    simp, },
+end
+
+/-- Auxiliary definition for `homotopy`. Use `homotopy.to_prev` instead. -/
+def to_prev' (f : Π i j, C.X i ⟶ D.X j) (i j : ι) : C.X i ⟶ D.X_prev j :=
+match c.prev j with
+| none := 0
+| some ⟨j',w⟩ := f i j' ≫ (D.X_prev_iso w).inv
+end
+
+lemma to_prev'_eq (f : Π i j, C.X i ⟶ D.X j) (i : ι) {j j' : ι} (w : c.rel j' j) :
+  to_prev' f i j = f i j' ≫ (D.X_prev_iso w).inv :=
+begin
+  dsimp [to_prev'],
+  rw c.prev_eq_some w,
+  refl,
+end
+
+@[simp] lemma to_prev'_zero (i j : ι) : to_prev' (λ i j, (0 : C.X i ⟶ D.X j)) i j = 0 :=
+begin
+  dsimp [to_prev'],
+  rcases c.prev j with ⟨⟩|⟨⟨j', w⟩⟩;
+  { dsimp [to_prev'._match_1], simp, },
+end
+
+@[simp] lemma to_prev'_add (f g : Π i j, C.X i ⟶ D.X j) (i j : ι) :
+  to_prev' (λ i j, f i j + g i j) i j = to_prev' f i j + to_prev' g i j :=
+begin
+  dsimp [to_prev'],
+  rcases c.prev j with _|⟨j',w⟩,
+  exact (zero_add _).symm,
+  exact preadditive.add_comp _ _ _ _ _ _,
+end
+
+@[simp] lemma to_prev'_neg (f : Π i j, C.X i ⟶ D.X j) (i j : ι) :
+  to_prev' (λ i j, -(f i j)) i j = - to_prev' f i j :=
+begin
+  dsimp [to_prev'],
+  rcases c.prev j with _|⟨j',w⟩,
+  exact neg_zero.symm,
+  exact preadditive.neg_comp _ _,
+end
+
+@[simp] lemma to_prev'_comp_left (f : C ⟶ D) (g : Π i j, D.X i ⟶ E.X j) (i j : ι) :
+  to_prev' (λ i j, f.f i ≫ g i j) i j = f.f i ≫ to_prev' g i j :=
+begin
+  dsimp [to_prev'],
+  rcases c.prev j with _|⟨j',w⟩,
+  { exact comp_zero.symm, },
+  { dsimp [to_prev'._match_1, hom.prev._match_1],
+    simp, },
+end
+
+@[simp] lemma to_prev'_comp_right (f : Π i j, C.X i ⟶ D.X j) (g : D ⟶ E) (i j : ι) :
+  to_prev' (λ i j, f i j ≫ g.f j) i j = to_prev' f i j ≫ g.prev j :=
+begin
+  dsimp [to_prev', hom.prev],
+  rcases c.prev j with _|⟨j',w⟩,
+  { exact zero_comp.symm, },
+  { dsimp [to_prev'._match_1, hom.prev._match_1],
+    simp, },
+end
+
+/--
+A homotopy `h` between chain maps `f` and `g` consists of components `h i j : C.X i ⟶ D.X i`
+which are zero unless `c.rel j i`,
+satisfying the homotopy condition.
+-/
+@[ext, nolint has_inhabited_instance]
+structure homotopy (f g : C ⟶ D) :=
+(hom : Π i j, C.X i ⟶ D.X j)
+(zero' : ∀ i j, ¬ c.rel j i → hom i j = 0 . obviously)
+(comm' : ∀ i,
+  f.f i = to_prev' hom i i ≫ D.d_to i + C.d_from i ≫ from_next' hom i i + g.f i . obviously')
+
+variables {f g}
+namespace homotopy
+
+restate_axiom homotopy.zero'
+
+/--
+The component of a homotopy from `next i` to `j`.
+-/
+def from_next (h : homotopy f g) (i j : ι) : C.X_next i ⟶ D.X j :=
+from_next' h.hom i j
+
+/--
+The component of a homotopy from `i` to `prev j`.
+-/
+def to_prev (h : homotopy f g) (i j : ι) : C.X i ⟶ D.X_prev j :=
+to_prev' h.hom i j
+
+lemma comm (h : homotopy f g) (i : ι) :
```

Comment:
Does it make sense to tag this with `@[simp]`? I'm not sure...

Reply:
This feels a bit scary to leave in `simp`s hands at this stage. It's certainly not simpler on the RHS, so it would require being pretty confident that there are actually proofs that can successfully employ a terminal `simp` using this equation --- and I don't think that's the case.

Reply:
The only situation I can imagine is if some cancellation can happen with the RHS of this equation. E.g. if the LHS appears in the middle of a composition with some differentials from the complex (which does show up in practice), so simp would be able to do the cancellation automatically. 

Anyway, I agree that it doesn't make sense to add `@[simp]` right now.